src/components/app/DragAndDrop.js

```diff
@@ -0,0 +1,222 @@
+|};
+
+type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
+
+/**
+ * Creates a target area to drop files on. Any elements which should
+ * be part of the target area should be wrapped by this component.
+ * A dropped file on this component or any of its children will be
+ * loaded into the profiler.
+ *
+ * By default, a <DragAndDropOverlay /> component will be added as a
+ * sibling to <DragAndDrop>, which will display a message asking the
+ * user to drop a file whenever a drag is detected. This overlay will
+ * cover the whole target area.
+ *
+ * <DragAndDrop>
+ *   ... children ...
+ * </DragAndDrop>
+ * <DragAndDropOverlay />  <-- added by default
+ *
+ * If a smaller overlay message is desired, <DragAndDropOverlay /> can
+ * be explicitly added as a child of <DragAndDrop>. That overlay will
+ * override the default one and be shown instead. This can give a
+ * nicer visual appearance for some views.
+ *
+ * <DragAndDrop>
+ *   ... children ...
+ *   <section>
+ *     <DragAndDropOverlay />  <-- manually added
+ *   <section>
+ * </DragAndDrop>
+ * (no default overlay added here anymore)
+ */
+
+class DragAndDropImpl extends React.PureComponent<Props> {
+  componentDidMount() {
+    // Prevent dropping files on the document.
+    document.addEventListener('drag', _dragPreventDefault, false);
+    document.addEventListener('dragover', _dragPreventDefault, false);
+    document.addEventListener('drop', _dragPreventDefault, false);
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('drag', _dragPreventDefault, false);
+    document.removeEventListener('dragover', _dragPreventDefault, false);
+    document.removeEventListener('drop', _dragPreventDefault, false);
+  }
+
+  _startDragging = (event: Event) => {
+    event.preventDefault();
+    this.props.startDragging();
+  };
+
+  _stopDragging = (event: Event) => {
+    event.preventDefault();
+    this.props.stopDragging();
+  };
+
+  _handleProfileDrop = (event: DragEvent) => {
+    event.preventDefault();
+    this.props.stopDragging();
+
+    if (!event.dataTransfer || !this.props.isNewProfileLoadAllowed) {
+      return;
+    }
+
+    const { files } = event.dataTransfer;
+    if (files.length > 0) {
+      this.props.retrieveProfileFromFile(files[0]);
+    }
+  };
+
+  render() {
+    const { className, children } = this.props;
+
+    return (
+      <>
+        <div
+          className={classNames(className, 'dragAndDropArea')}
+          onDragEnter={this._startDragging}
+          onDragExit={this._stopDragging}
+          onDrop={this._handleProfileDrop}
+        >
+          {children}
+        </div>
+        {/* Put the default overlay here if it is to be used. The
+          dragAndDropArea div creates its own stacking context, so
+          even if it contains children with high z-indexes, the
+          default overlay will still appear on top when shown.*/
+        this.props.useDefaultOverlay && <DragAndDropOverlay isDefault={true} />}
```

Comment:
Please use a ternary here instead of the `&&` switching. I find the `&&` and `||` behaviors take a lot more time to reason with what actually gets passed in, rather than the arm branching of a ternary.

`useDefaultOverlay ? <DragAndDropOverlay isDefault={true} />} : null`